### PR TITLE
Enforced -querier.max-query-lookback in the query-frontend for range queries

### PR DESCRIPTION
**Pull request by pracucci**
**What this PR does**:
In the PR #3452 I've introduced `-querier.max-query-lookback` but I've intentionally skipped enforcing the limit in the query-frontend, in order to keep the change  set smaller. In this PR I'm enforcing it in the query-frontend too, for range queries.

The main benefit of enforcing it in the query-frontend too is to reduce the number of query range requests enqueued for a tenant, if such queries cover a range outside the max lookback. Without this PR, the limit would be enforced anyway in the querier, but we may end up with a large number of splitted queries in the query-frontend/query-scheduler queue which would be noop once executed in the querier.

**Which issue(s) this PR fixes**:
N/A

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/util/validation/limits.go [hunk 1]
```diff
@@ -123,8 +123,8 @@ func (l *Limits) RegisterFlags(f *flag.FlagSet) {
...
 	f.DurationVar(&l.MaxQueryLength, "store.max-query-length", 0, "Limit the query time range (end - start time). This limit is enforced in the query-frontend (on the received query), in the querier (on the query possibly split by the query-frontend) and in the chunks storage. 0 to disable.")
-	f.DurationVar(&l.MaxQueryLookback, "querier.max-query-lookback", 0, "Limit how long back data (series and metadata) can be queried, up until <lookback> duration ago. 0 to disable.")
-	f.IntVar(&l.MaxQueryParallelism, "querier.max-query-parallelism", 14, "Maximum number of queries will be scheduled in parallel by the frontend.")
+	f.DurationVar(&l.MaxQueryLookback, "querier.max-query-lookback", 0, "Limit how long back data (series and metadata) can be queried, up until <lookback> duration ago. This limit is enforced in the query-frontend, querier and ruler. If the requested time range is outside the allowed range, the request will not fail but will be manipulated to only query data within the allowed time range. 0 to disable.")
+	f.IntVar(&l.MaxQueryParallelism, "querier.max-query-parallelism", 14, "Maximum number of splitted queries will be scheduled in parallel by the frontend.")
```
Comment (pstibrany) [inline pkg/util/validation/limits.go:127]: Just "split".

File: pkg/querier/querier.go [hunk 2]
```diff
@@ -512,7 +512,8 @@ func validateQueryTimeRange(ctx context.Context, userID string, startMs, endMs i
 		// Make sure to log it in traces to ease debugging.
 		level.Debug(spanlogger.FromContext(ctx)).Log(
 			"msg", "the end time of the query has been manipulated because of the 'max query into future' setting",
-			"original", origEndTime, "updated", endTime)
+			"original", origEndTime.String(),
```
Comment (pstibrany) [inline pkg/querier/querier.go:515]: Adding date-time formatting may be useful as well. WDYT? (Everywhere)

File: pkg/util/validation/limits.go [hunk 3]
```diff
@@ -330,7 +331,7 @@ func (o *Overrides) MaxQueriersPerUser(userID string) int {
 	return o.getOverridesForUser(userID).MaxQueriersPerTenant
 }
 
-// MaxQueryParallelism returns the limit to the number of sub-queries the
+// MaxQueryParallelism returns the limit to the number of splitted queries the
```
Comment (pstibrany) [inline pkg/util/validation/limits.go:334]: split

File: pkg/querier/queryrange/limits_test.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+	tests := map[string]struct {
+		maxQueryLookback  time.Duration
+		reqStartTime      time.Time
+		reqEndTime        time.Time
+		expectedSkipped   bool
+		expectedStartTime time.Time
+		expectedEndTime   time.Time
+	}{
+		"should not manipulate time range it max lookback is disabled": {
```
Comment (pstibrany) [inline pkg/querier/queryrange/limits_test.go:31]: typo: `it` -> `if`

File: pkg/querier/queryrange/limits_test.go [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+		},
+		"should fail on a query on large time range over the limit, ending now": {
+			maxQueryLength: thirtyDays,
+			reqStartTime:   now.Add(-thirtyDays).Add(-100 * time.Hour),
+			reqEndTime:     now,
+			expectedErr:    "the query time range exceeds the limit",
+		},
+		"should fail on a query on large time range over the limit, ending in the past": {
+			maxQueryLength: thirtyDays,
+			reqStartTime:   now.Add(-2 * thirtyDays).Add(-thirtyDays).Add(-100 * time.Hour),
```
Comment (pstibrany) [inline pkg/querier/queryrange/limits_test.go:146]: This is getting hard to read. What about this?

```suggestion
			reqStartTime:   now.Add(-3*thirtyDays - 100*time.Hour),
```

**Review by pstibrany — APPROVED** [review]
Nice. I especially like how you always leave the code in a better shape with each PR!

**Comment by pracucci** [comment]
Thanks @pstibrany for your thoughtful review. I've addressed all comments!

**Review by yeya24 — APPROVED** [review]
I don't know whether I can approve or not, but this looks nice!

**Comment by pracucci** [comment]
> I don't know whether I can approve or not, but this looks nice!

Sure you can!

**Review by ranton256 — APPROVED** [review]

**Pull request merged**